consensus/core/src/block_verifier.rs

```diff
@@ -1,31 +1,219 @@
+    fn verify_metadata(&self, block: &SignedBlock) -> ConsensusResult<()>;
+
+    /// Verifies a block's metadata.
+    fn verify_transactions(&self, block: &SignedBlock) -> ConsensusResult<()>;
+
+    /// Verifies a block's metadata.
+    // TODO: see if it makes sense to take &SignedBlock as input.
+    fn verify_ancestors(&self, block: &VerifiedBlock) -> ConsensusResult<()>;
+}
+
+/// `SignedBlockVerifier` checks the validity of a block.
+///
+/// Blocks that fail verification at one honest authority will be rejected by all other honest
+/// authorities as well. The means invalid blocks, and blocks with an invalid ancestor, will never
+/// be accepted into the DAG.
+///
+/// Verification methods can be called from different components and threads.
+pub(crate) struct SignedBlockVerifier {
+    context: Arc<Context>,
+    transaction_verifier: Arc<dyn TransactionVerifier>,
+}
+
+impl SignedBlockVerifier {
+    pub(crate) fn new(
+        context: Arc<Context>,
+        transaction_verifier: Arc<dyn TransactionVerifier>,
+    ) -> Self {
+        Self {
+            context,
+            transaction_verifier,
+        }
+    }
+
+    fn verify_signature(&self, block: &SignedBlock) -> ConsensusResult<()> {
+        let committee = &self.context.committee;
+        ensure!(
+            committee.is_valid_index(block.author()),
+            ConsensusError::InvalidAuthorityIndex {
+                index: block.author(),
+                max: self.context.committee.size() - 1
+            }
+        );
+        let authority = committee.authority(block.author());
+
+        let signature = ProtocolKeySignature::from_bytes(block.signature())?;
+
+        let inner_digest = block.compute_inner_digest()?;
+        signature
+            .verify(
+                &to_consensus_block_intent(inner_digest),
+                &authority.protocol_key,
+            )
+            .map_err(ConsensusError::SignatureVerificationFailure)
+    }
 }
 
-#[derive(Clone)]
-pub struct TestBlockVerifier;
+// All block verification logic are implemented below.
+impl BlockVerifier for SignedBlockVerifier {
+    fn verify_metadata(&self, block: &SignedBlock) -> ConsensusResult<()> {
+        let committee = &self.context.committee;
+        // The block must belong to the current epoch and have valid authority index,
+        // before having its signature verified.
+        if block.epoch() != committee.epoch() {
+            return Err(ConsensusError::WrongEpoch {
+                expected: committee.epoch(),
+                actual: block.epoch(),
+            });
+        }
+        if !committee.is_valid_index(block.author()) {
+            return Err(ConsensusError::InvalidAuthorityIndex {
+                index: block.author(),
+                max: committee.size() - 1,
+            });
+        }
+        // Verifiy the block's signature.
+        self.verify_signature(block)?;
+        // Verify the block's ancestor refs is consistent with the round of the block,
```

Comment:
why is this not in verify_ancestors()?

Reply:
This does not require block content. In general it is better to run verifications as soon as the necessary data are available.